### PR TITLE
Cli mode on Windows and Folder search

### PR DESCRIPTION
**Pull request by thiagoalz**
Adding support do windows on Cli mode and searching subtitles for all videos in a folder.

Its a good idea to check if it is still working on Linux and mac.
(it should)

### CODE REVIEW

**Comment by emericg** [comment]
Hi!
The three different features you got there (wget/gzip bypass, windows support and search by folder) are pretty neat, and they were all on my todo list!
I'll need a bit of time to test all of that on mac and linux and python 2 and 3, and I may ask a few adjustments and to separate the features, but overall that seems awesome, thanks for your work ;-)

**Comment by thiagoalz** [comment]
You are welcome! :D 
Just let me know if you need something.

I`ll try to find sometime to remove gzip e cat from ther rest of the code and test on linux too.

File: OpenSubtitlesDownload.py [hunk 1]
```diff
@@ -523,7 +523,13 @@ def dependencyChecker():
     # Go through the paths taken from arguments, and extract only valid video paths
     for i in result.filePathListArg:
         filePath = os.path.abspath(i)
-        if checkFileValidity(filePath):
+        if os.path.isdir(filePath): #If it is a dir, get all files recursively
```
Comment (emericg) [inline OpenSubtitlesDownload.py]: As a general rule, follow same coding/comments style from the upstream project ^^ (also, be careful about trailing whitespaces)

File: OpenSubtitlesDownload.py [hunk 2]
```diff
@@ -523,7 +523,13 @@ def dependencyChecker():
...
+            for root, dirs, files in os.walk(filePath, topdown=False):
```
Comment (emericg) [inline OpenSubtitlesDownload.py]: The "dirs" option should be removed. One level is plenty. It's extra dangerous depending on where you launch the script, I now some folders on my computer that would spawn thousands of instances of the scripts, with no way of stopping the carnage!

File: OpenSubtitlesDownload.py [hunk 3]
```diff
@@ -585,7 +592,7 @@ def dependencyChecker():
         process_videoDispatched = subprocess.Popen(command_splitted)
 
     # Do not spawn too many instances at the same time
-    time.sleep(0.33)
+    time.sleep(0.5)
```
Comment (emericg) [inline OpenSubtitlesDownload.py]: why?

File: OpenSubtitlesDownload.py [hunk 4]
```diff
@@ -571,7 +577,8 @@ def dependencyChecker():
     if not (len(opt_languages) == 1 and opt_languages[0] == 'eng'):
         for resultlangs in opt_languages:
             command += " -l " + resultlangs
-
+    command = "python " + command;
```
Comment (emericg) [inline OpenSubtitlesDownload.py]: Most definitly not. That will only work on your computer.
see the "# Get OpenSubtitlesDownload.py script path" section and add necessary code there if the autodetection is not working for you, but that should really be a separate patch.

File: OpenSubtitlesDownload.py [hunk 5]
```diff
@@ -88,10 +88,10 @@
 # - manual (always let you choose the subtitles you want)
 # - default (in case of multiple results, let you choose the subtitles you want)
 # - auto (automatically select the best subtitles found)
-opt_selection_mode = 'default'
+opt_selection_mode = 'auto'
```
[the remark this answers is not in the document]
Comment (thiagoalz) [reply OpenSubtitlesDownload.py]: Hum... Basically, since the default search is "hash than filename", for all my tests , when I had 2 or more options they were the same. So having to chose was the exception.
But no problem, I dont have any research to see what most people need. Just my case/feeling :)

Probably your option is as good!

File: OpenSubtitlesDownload.py (hunk 1 above)
Comment (thiagoalz) [reply OpenSubtitlesDownload.py]: Agreed :)

File: OpenSubtitlesDownload.py (hunk 2 above)
Comment (thiagoalz) [reply OpenSubtitlesDownload.py]: Actualy not really true... The file is only added to the vector if it is a video (checkFileValidity) and does not already have a subtitle (checkSubtitlesExists).
Only after this the script instances are created.

I just tested the script running in my c:\windows folder and it worked perfectly :)

File: OpenSubtitlesDownload.py (hunk 3 above)
Comment (thiagoalz) [reply OpenSubtitlesDownload.py]: Maybe because I has testing too much, I started to receive service denials from OpenSubTitles.
After I increased the delay a little bit, the denials stoped.

File: OpenSubtitlesDownload.py (hunk 4 above)
Comment (thiagoalz) [reply OpenSubtitlesDownload.py]: I Think you didnt get the point here....
The script path is working, however I could run a python script like this:
[1]./PathToScript/Script.py
or this?
[2]python /PathToScript/Script.py

Im not sure why, my in my wondows install the option [1] is not working.
So i changes to the option [2] that should work for everyone ele too.
This fix was part to windows compatibility fix.

File: OpenSubtitlesDownload.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (thiagoalz) [reply OpenSubtitlesDownload.py]: Created a config File so we can have separate discussions.

File: OpenSubtitlesDownload.py (hunk 2 above)
Comment (emericg) [reply OpenSubtitlesDownload.py]: If you run the script with that option, it sure works, but every single file on the c:/windows folder will be probed by the script, and every video file will be picked up and an instance will be launched. If an user launch this on its movie folder, he's in for a surprise!

File: OpenSubtitlesDownload.py (hunk 3 above)
Comment (emericg) [reply OpenSubtitlesDownload.py]: Ok, I will keep an eye for this if we start to see some more errors.

File: OpenSubtitlesDownload.py (hunk 4 above)
Comment (emericg) [reply OpenSubtitlesDownload.py]: [1] tells your shell to use the interpreter specified by the shebang, the first line of the script. We have /usr/bin/env python set there, which is the most portable way to do that. https://docs.python.org/3/using/windows.html#shebang-lines

[2] tells your shell to use "python" binary from your path to start the script. You can start the script using that line, but you cannot set an hardcoded interpreter inside the script. The line you modified is supposed to launch secondary instance of the script with the same interpreter used to launch the first instance.

I did get the point, but someone might launch the script with /user/roger/python3.6, and secondary instances are expected to use the same python interpreter. And even especially on windows, people might install python but not setting it up in the system's path, so respecting the path used to launch the script is important.
You need to see why os.path.isabs(sys.argv[0]) do not get the absolute interpreter path. Even if you use [1], it works on linux, BSD and macOS.

**Comment by emericg** [comment]
Sorry for the delay I wasn't home for the past few days.

I can commit modified versions of some of your patches if you want, namely:
- (parts of) Adapting to work on windows in CLI Mode & Addind windows support on readme file
- Adding support do download subtitles from all files in folder
- Removing dependencies on Cli mode & ￼Fix for python 2 compatibility
- Adding --force option to overwrite subtitles

Commit authoring will be kept, and tell me if you want your name and email in the contributors list.

The ini file is nice, it's what we did in https://github.com/emericg/OpenSubtitlesDownloadQt, but it goes against the goal of this one: one file, very limited dependencies (that's why removing wget and gunzip is great for the CLI).

**Comment by thiagoalz** [comment]
Sure... go ahead.
Can you get the commit directly or should I create a new pull request?

File: OpenSubtitlesDownload.py (hunk 4 above)
Comment (thiagoalz) [reply OpenSubtitlesDownload.py]: The problem with shebang is that it works only on Unix systems.

I just changed the script to use sys.executable instead of "python".
Now it should keep the same interpreter and stay compatible with unix and windows.

(Without this, it just does not work on windows)

**Comment by emericg** [comment]
OK commits have been squashed and merged, thanks!

The "Adding --force option to overwrite subtitles" will be commited soon but in inverted, a flag to "not overwrite".
The overwrite will stay on by default, its how the software has always worked and with the GUI it doesn't make sense to invert it. How I typically use it: click on a file, search subtitles, check the result, if subtitles not synced, retry. Having to delete the subtitles manually each time, and/or having to explain to user why the software doesn't even start if a subtitles is already there doesn't really make sense.

And by the way you where right about the usage of sys.executable to create new instances, it was indeed needed on windows (and also actually hiding a bug on unixes platforms too...).
